---
title: "Normalise scRNA-seq gene expression data"
output: html_notebook
---

Nuha BinTayyash, 2020

This notebook shows how run [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) R package to normalize ScRNA-seq gene expression data for highly expressed genes in Islet $\alpha$ cell from [GSE8737 single cell RNA-seq ](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87375) dataset.

Load ScRNA-seq gene expression data for highly expressed genes in Islet $\alpha$ cell from [GSE8737 single cell RNA-seq ](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87375) dataset. 

\* the dataset is large you can download it from the [GSE8737 single cell RNA-seq ](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87375)
```{r}
counts <- read.csv(file = 'GSE87375_Single_Cell_RNA-seq_Gene_Read_Count.csv',row.names = 1, header = TRUE)
dim(counts)
```
Drop ERCC spike-in 
```{r}
counts<-counts[grep("ENSMUSG", rownames(counts)), ]
```
# Scale  ð›¼  pseudotime information
```{r}
alpha_pseudotime  <- read.csv(file = '../data/alpha_pca.csv',row.names = 1, header = TRUE)
X <- alpha_pseudotime['Pseudotime']
X_scaled_alpha <- (X - min(X)) / (max(X) - min(X)) 
write.csv(X_scaled_alpha , file = '../data/alpha_time_points.csv')

```

Load  ð›¼  pseudotime information
```{r}
alpha_col_data <- read.csv(file = '../data/alpha_time_points.csv',row.names = 1, header = TRUE)
head(alpha_col_data)
```
Get gene expression data for  ð›¼  cells
```{r}
alpha_counts <- counts[ , grepl( "a" , names( counts ) ) ]
alpha_counts <- alpha_counts[rownames(alpha_col_data)]
keep <- rowMeans(alpha_counts) >.1
alpha_counts <- alpha_counts[keep,]
dim(alpha_counts)
```

## Normalize gene expression data using DESeq2
```{r}
library("DESeq2")
dds <- DESeqDataSetFromMatrix(countData = alpha_counts,
                              colData = alpha_col_data,
                              design = ~pseudotime)
dds <- estimateSizeFactors(dds)
normalized_alpha_counts <-counts(dds, normalized=TRUE)
dim(normalized_alpha_counts)
write.csv(normalized_alpha_counts, file = "../data/normalized_alpha_counts.csv")
```

## Run one sample test
```{r}
dds <- DESeq(dds, test="LRT", reduced = ~ 1)
res <- results(dds)
dim(as.data.frame(res))
write.csv(as.data.frame(res),file="../data/alpha_DESeq2.csv")
sessionInfo()
```
